pyro/infer/reparam/stable.py

```diff
@@ -99,11 +99,122 @@ def __call__(self, name, fn, obs):
+        # 1. WLOG scale and shift so s=1 and m=0, additionally shifting to convert
+        #    from Zolotarev's S parameterization to Nolan's S0 parameterization.
+        # 2. Decompose X = S + T, where
+        #    S ~ S(a,0,...,0) is symmetric and
+        #    T ~ S(a,sgn(b),...,0) is totally skewed.
+        # 3. Decompose S = G * sqrt(Z) via the symmetric strategy, where
+        #    Z ~ S(a/2,1,...,0) is totally-skewed and
+        #    G ~ Normal(0,1) is Gaussian.
+        # 4. Defer the totally-skewed Z and T to the Chambers-Mallows-Stuck
+        #    strategy: Z = f(Unif,Exp), T = f(Unif,Exp).
+        #
+        # To derive the parameters of S and T, we solve the equations
+        #
+        #   T.stability = a            S.stability = a
+        #   T.skew = sgn(b)            S.skew = 0
+        #   T.loc = 0                  S.loc = 0
+        #
+        #   s = (S.scale**a + T.scale**a)**(1/a) = 1       # by step 1.
+        #
+        #       S.skew * S.scale**a + T.skew * T.scale**a
+        #   b = ----------------------------------------- = sgn(b) * T.scale**a
+        #                S.scale**a + T.scale**a
+        # yielding
+        #
+        #   T.scale = |b| ** (1/a)     S.scale = (1 - |b|) ** (1/a)
+
+        # Draw parameter-free noise.
+        proto = fn.stability
+        half_pi = proto.new_full(proto.shape, math.pi / 2)
+        one = proto.new_ones(proto.shape)
+        zu = pyro.sample("{}_z_uniform".format(name),
+                         self._wrap(dist.Uniform(-half_pi, half_pi), event_dim))
+        ze = pyro.sample("{}_z_exponential".format(name),
+                         self._wrap(dist.Exponential(one), event_dim))
+        tu = pyro.sample("{}_t_uniform".format(name),
+                         self._wrap(dist.Uniform(-half_pi, half_pi), event_dim))
+        te = pyro.sample("{}_t_exponential".format(name),
+                         self._wrap(dist.Exponential(one), event_dim))
+
+        # Differentiably transform.
+        a = fn.stability
+        z = _unsafe_standard_stable(a / 2, 1, zu, ze, coords="S")
+        t = _standard_stable(a, one, tu, te, coords="S0")
+        a_inv = a.reciprocal()
+        skew_abs = fn.skew.abs()
+        t_scale = skew_abs.pow(a_inv)
+        s_scale = (1 - skew_abs).pow(a_inv)
+        shift = _safe_shift(a, skew_abs, t_scale)
+        loc = fn.loc + fn.scale * fn.skew.sign() * (t * t_scale + shift)
+        scale = fn.scale * s_scale * z.sqrt() * (math.pi / 4 * a).cos().pow(a_inv)
+        scale = scale.clamp(min=torch.finfo(scale.dtype).tiny)
+
+        # Construct a scaled Gaussian, using Stable(2,0,s,m) == Normal(m,s*sqrt(2)).
+        new_fn = self._wrap(dist.Normal(loc, scale * (2 ** 0.5)), event_dim)
+        return new_fn, obs
+
+
+def _unsafe_shift(a, skew_abs, t_scale):
+    # At a=1 the lhs has a root and the rhs has an asymptote.
+    return (t_scale - skew_abs) * (math.pi / 2 * a).tan()
```

Comment:
I think it will be easier to understand the cancelation and patching logic here if you remove `t_scale` arg.

Reply:
Ah, I see, you want to save computation when no hole happens.